src/infrastructure/arranging/ordering.rs

```diff
@@ -0,0 +1,441 @@
+    ///
+    /// It is important to note that although the returned stream has the capacity specified,
+    /// the stream will have a zero length.
+    ///
+    /// [`OrderingStream`]: ./struct.OrderingStream.html
+    pub fn with_capacity(size: usize, stream_id: u8) -> OrderingStream<T> {
+        OrderingStream {
+            storage: HashMap::with_capacity(size),
+            expected_index: 1,
+            stream_id,
+        }
+    }
+
+    /// Returns the identifier of this stream.
+    fn stream_id(&self) -> u8 {
+        self.stream_id
+    }
+
+    /// Returns the next expected index.
+    pub fn expected_index(&self) -> usize {
+        self.expected_index
+    }
+
+    /// Returns an iterator of stored items.
+    ///
+    /// # Algorithm for returning items from an Iterator.
+    ///
+    /// 1. See if there is an item matching our `expected_index`
+    /// 2. If there is return the `Some(item)`
+    ///    - Increase the `expected_index`
+    ///    - Start at '1'
+    /// 3. If there isn't return `None`
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// let stream = OrderingStream::new();
+    ///
+    /// let iter = stream.iter_mut();
+    ///
+    /// while let Some(item) = iter.next() {
+    ///    // Items from iterator are in order.
+    /// }
+    /// ```
+    ///
+    /// # Remarks
+    /// - Iterator mutates the `expected_index`.
+    /// - You can't use this iterator for iterating trough all cached values.
+    pub fn iter_mut(&mut self) -> IterMut<T> {
+        IterMut {
+            items: &mut self.storage,
+            expected_index: &mut self.expected_index,
+        }
+    }
+}
+
+impl<T> Arranging for OrderingStream<T> {
+    type ArrangingItem = T;
+
+    /// Will order the given item based on the ordering algorithm.
+    ///
+    /// With every ordering operation an `incoming_index` is given. We also keep a local record of the `expected_index`.
+    ///
+    /// # Algorithm
+    ///
+    /// There are three scenarios that are important to us.
+    /// 1. `incoming_index` == `expected_index`.
+    /// This package meets the expected order, so we can return it immediately.
+    /// 2. `incoming_index` > `expected_index`.
+    /// This package is newer than we expect, so we have to hold it temporarily until we have received all previous packages.
+    /// 3. `incoming_index` < `expected_index`
+    /// This can only happen in cases where we have a duplicated package. Again we don't give anything back.
+    ///
+    /// # Remark
+    /// - When we receive an item there is a possibility that a gab is filled and one or more items will could be returned.
+    ///   You should use the `iter_mut` instead for reading the items in order.
+    ///   However the item given to `arrange` will be returned directly when it matches the `expected_index`.
+    fn arrange(
+        &mut self,
+        incoming_offset: usize,
+        index: Self::ArrangingItem,
```

Comment:
This is likely a typo. I suspect you wanted `item` here rather than index.

Reply:
Good catch!